ECMWF Open Data download and conversion functions

swarnaleem opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->
ECMWF Open Data 15-day forecast download and conversion to NetCDF (from Grib2) 
## Description
<!--- Describe your changes in detail -->
The download.ECMWF.R function downloads ECMWF Opendata 15-day perturbed forecast (`pf`) consisting of 50 ensemble members using `ecmwf-opendata` python package. Currently, it is downloading 50 ensemble members using `download_ecmwf_opendata.py`. The parameters being downloaded are as follows:
Parameter Short Name | Name
-- | --
2t | 2 metre temperature
tp | total precipitation
10u | 10 metre U wind component
10v | 10 metre V wind component
q | Specific humidity
r | Relative humidity
sp | Surface pressure

After downloading, `ecmwf_grib2nc.py` function converts the `grib2` file consisting of 50 ensemble members into 50 separate `netCDF` files based on their ensemble numbers.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

PEcAn currently has 17 meteorological drivers available, some of which are Ameriflux, ERA5, and CMIP5 which are used as input data to various models. ECMWF Open Data provides globally available ensemble forecasts that can be used to drive various models (available in PEcAn).

Creating an ECMWF Open Data pipeline in PEcAn would help in adding various meteorological forecast data (weather data) as input variables to the ecosystem models which would help in running them in the future. In this proposal, the interest is around 15-day ensemble forecast that originates from the day of the forecast and goes 360h into the future. 

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [x] Within one week
- [ ] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
WIP ECMWF opendata download and conversion

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+#' Download ECMWF Open Data 15 day forecast data products
+#'
```
please document all function args, and export the fcn, e.g.

```
##' @export
##' @param outfolder location on disk where outputs will be stored
##' @param fchour
```

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+#' @author Swarnalee Mazumder
+#'
+
+library(reticulate)
```
please move library calls to `DESCRIPTION` file, e.g. see [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.remote/DESCRIPTION#L26). I guess in this case you'll need to move `reticulate` from `Suggests` to `Imports` for `data.atmosphere`

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+  
+  all_parameters <- c("2t", "tp", "10u", "10v", "q", "r", "sp")
+  
+  if (tolower(parameters) == "all") {
```
if this condition is not met there will be no `parameter_types`, then line 61 will most likely throw a `'parameter_types' not found` error, should there be an `else` to this `if`?

swarnalee-m replied:
Added `else` to show what can be given as an input, here `all`.

istfer replied:
I believe this is remains to be the case, please add an `else` to make sure `parameter_types` are assigned something in case `tolower(parameters) != "all"`

swarnalee-m replied:
If `tolower(parameters) != "all"` should the `parameter_types` still be `"all"`?

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+    time <- 12
+  }
+  
+  all_parameters <- c("2t", "tp", "10u", "10v", "q", "r", "sp")
```
really a pity that no radiation variable is available in the open dataset, commenting here to remember follow up on this

istfer replied:
they will probably extend the open dataset to include more ENS parameters later, currently it seems clear that we don't have radiation closing this now

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+  if (stream == "enfo"){
+    time <- time
+    step <- 360
+    type <- type
```
just as a reminder, as we discussed we want `cf` too

swarnalee-m replied:
made changes to accommodate for `cf`

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+    type <- type
+  }  else{ 
+    PEcAn.logger::logger.severe(sprintf(
+      "Invalid data stream."
```
Could you extend the message to mention the available streams (currently only `enfo`)

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+  }
+  
+  # Python script "download_ecmwf.py" to get latest forecast date and download forecast datasets
+  source_python("download_ecmwf.py")
```
could you make sure we can successfully locate this file, e.g. see [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.remote/inst/download_SMAP_gee2pecan.R#L63)

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+  
+  # Python script "ecmwf_grib2nc.py" to convert grib2 to netCDF
+  # Converting Grib2 file to ensemble wise NetCDF file
+  source_python("ecmwf_grib2nc.py")
```
same as above

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,129 @@
+  # Converting Grib2 file to ensemble wise NetCDF file
+  source_python("ecmwf_grib2nc.py")
+  nc_ecmwf <- grib2nc_ecmwf(grib_fname)
+  
```
`download.*.R` functions in `data.atmosphere` module are supposed to return a results data frame if I'm not mistaken, e.g. see [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.atmosphere/R/download.AmerifluxLBL.R#L185) and [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.atmosphere/R/download.GFDL.R#L54)

istfer reviewed (changes requested):
Thanks Swarnalee for this WIP PR, I started a review (not completed yet). In addition to those comments, could you also update the documentation for ECMWF? I believe [this](https://github.com/PecanProject/pecan/blob/develop/book_source/03_topical_pages/06_data/01_meteorology.Rmd) is the right place to do it

Commit:
*added ability to access point data
*modified datasets acc to CF standard
*updated documentation
*resolved PR comments

istfer reviewed (changes requested):
I'll continue to review the revised code but in the meantime could you please make sure this PR would make ECMWF pipeline run successfully through the workflow. I still don't see a registration file, which should look something like [this](https://github.com/PecanProject/pecan/blob/develop/modules/data.atmosphere/inst/registration/register.NOAA_GEFS.xml). 

I also suspect that `fchour` will never be passed to `download.ECMWF `by upstream functions, i.e. see [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.atmosphere/R/download.raw.met.module.R#L99). You can either change that fcn arg to `product` (logically it's not that far off I'd be fine with it) or you need to add `fchour` to that `convert.input `call

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+#'   the same target name (i.e. same variable) that already exist in
+#'   `outfolder`. If `TRUE`, silently overwrite existing files.
+#' @return information about the output file
+#' @export
```
Could you also run roxygen (package v7.1.2) so that the `NAMESPACE` will be updated

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+  }
+  
+  # Python script "download_ecmwf.py" to get latest forecast date and download forecast datasets
+  script.path = file.path(system.file("download_ecmwf_opendata.py", package = "PEcAn.data.atmosphere"))
```
```suggestion
  script.path = file.path(system.file("ECMWF/download_ecmwf_opendata.py", package = "PEcAn.data.atmosphere"))
```

istfer committed:
Update modules/data.atmosphere/R/download.ECMWF.R

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+  
+  # Python script "ecmwf_grib2nc.py" to convert grib2 to netCDF
+  # Converting Grib2 file to ensemble wise NetCDF file
+  script.path = file.path(system.file("ecmwf_grib2nc.py", package = "PEcAn.data.atmosphere"))
```
```suggestion
  script.path = file.path(system.file("ECMWF/ecmwf_grib2nc.py", package = "PEcAn.data.atmosphere"))
```

istfer committed:
Update modules/data.atmosphere/R/download.ECMWF.R

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+                           lat.in,
+                           lon.in,
+                           fchour, 
+                           stream = "enfo", 
```
```suggestion
                           product = NULL, 
```

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+    dir.create(outfolder, showWarnings = FALSE, recursive = TRUE)
+  }
+  
+  
```
```suggestion
  if(is.null(product)) product <- "enfo"
```

istfer replied:
as we decided `product` to distinguish between `open-data` and `s2s` (pending on the decision whether we want this dataset), I'm resolving this too. `download.ECMWF` function will continue to have `stream="enfo"`, `fchour=00` arguments and will in addition have `product="open-data"`

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+download.ECMWF <- function(outfolder, 
+                           lat.in,
+                           lon.in,
+                           fchour, 
```
```suggestion
                           fchour = 00, 
```

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+#'
+#' @param outfolder location on disk where outputs will be stored
+#' @param lat.in,lon.in site coordinates, decimal degrees (numeric)
+#' @param fchour reference time of the forecasts. Values are 00, 12 for 15 day forecast.
```
document that fchour is not passed via upstream functions

istfer commented on book_source/03_topical_pages/06_data/01_meteorology.Rmd:
```diff
@@ -132,3 +132,11 @@ Resolution: 30 min
+
+Scale: Global
+
+Resolution: 6 hour, 0.4 degree
```
document that current implementation only uses forecast hour 00 when download function is called from within the PEcAn workflow

istfer commented on book_source/03_topical_pages/06_data/01_meteorology.Rmd:
```diff
@@ -132,3 +132,11 @@ Resolution: 30 min
+
+Resolution: 6 hour, 0.4 degree
+
+Availability: Varies by [day](https://data.ecmwf.int/)
```
document that 15 day forecast is provided at two hours 00 and 12, but  current implementation only uses forecast hour 00 when download function is called from within the PEcAn workflow

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+  }
+  
+  # Removes any already existing files with base name same as `latest_filedate``
+  if ((length(list.files(path = outfolder, pattern = as.character(latest_filedate)) ) > 0) == TRUE){
```
do we really want to remove or if the global files are already there  can  we skip the download?

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+    file.remove(list.files(outfolder, pattern = as.character(latest_filedate)))
+  }
+  
+  fname <- paste(latest_filedate, time, step, stream, type, sep = "_")
```
when `type = c("cf", "pf")`, `fname` and subsequently `in_filename` will also be a vector right? e.g. "`20220727_0_360_enfo_cf.grib2`" and "`20220727_0_360_enfo_pf.grib2`". Is `ecmwfdownload` expecting a vector for the `in_filename` arg?

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+  
+  fname <- paste(latest_filedate, time, step, stream, type, sep = "_")
+  
+  in_filename <- paste0(fname, ".grib2")
```
in_filename contains only a file name, then ecmwfdownload downloads the file wherever the working directory is. Please chage it to include the full path
```suggestion
  in_filename <- file.path(outfolder, paste0(fname, ".grib2"))
```

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,201 @@
+  script.path = file.path(system.file("ECMWF/ecmwf_grib2nc.py", package = "PEcAn.data.atmosphere"))
+  reticulate::source_python(script.path)
+  
+  nc_ecmwf <- grib2nc_ecmwf(in_filename, outfolder, out_filename)
```
`out_filename` is not specified anywhere in this function

istfer replied:
in addition, looking at `grib2nc_ecmwf` this function call is missing 2 required args '`lat_in`' and '`lon_in`'

istfer reviewed (changes requested):
please fix the issues regarding the files/filenames

swarnaleem committed:
Update modules/data.atmosphere/R/download.ECMWF.R

Co-authored-by: istfer <fer.istem@gmail.com>

Commit:
* resolved comments
* updated documentation

Commit:
downloads 15day forecast files

Commit:
updated ecmwf download function added met2CF

istfer commented on modules/data.atmosphere/R/ECMWF_helper_functions.R:
```diff
@@ -0,0 +1,283 @@
+  
+  ##### 3h CF -> 0 - 141h at 3h timestep
+  cf_3h_out_fname <- paste0(latest_filedate, time, stream, "cf_3h", sep = "_")
+  cf_3h_nc <- combine_files_cf(fnames_3h, lat_in, lon_in, out_filename= cf_3h_out_fname)
```
I thought we agreed to not write these intermediate 3hrly and 6 hrly netcdf files, are these still being written?

Commit:
removed creation of intermediate cf pf netCDF files

istfer committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

Commit:
changed function name

Commit:
registration file

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,242 @@
+  script.path = file.path(system.file("ECMWF/download_ecmwf.py", package = "PEcAn.data.atmosphere"))
+  reticulate::source_python(script.path)
+  
+  date_latestdata <- ecmwflatest(time, step, stream, type, all_parameters)
```
```suggestion
  date_latestdata <- ecmwflatest(time, step_15day, stream, type, all_parameters)
```
there is not `step` variable, was this supposed to be `step_15day`?

istfer commented on modules/data.atmosphere/R/download.ECMWF.R:
```diff
@@ -0,0 +1,242 @@
+    
+    in_fname <- paste(latest_filedate, time, stream, sep = "_")
+    
+    data_download <- ecmwfdownload(date, time, stream, type, all_parameters, in_fname)
```
I haven't thoroughly tested, but this function seems to be downloading the grib files to the working directory whereas it probably should use `outfolder`

istfer committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

infotroph committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

mdietze commented:
@swarnalee-m pinging you again to hopefully finish up this PR

infotroph committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

infotroph committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

mdietze committed:
Merge branch 'develop' into ecmwf-opendata-download-conversion

mdietze commented:
@istfer any chance you can finish up this PR? @swarnaleem hasn't touched it in a year and doesn't seem to be responding to inquiries about addressing the changes you requested.

istfer commented:
I can give it a try

infotroph commented:
I propose we close this PR -- ECMWF has reorganized their API and no longer provides the 0.4 degree beta product (see https://www.ecmwf.int/en/forecasts/datasets/open-data, especially cmd-F "beta" in the product description section), so getting the existing code working will now be a substantial lift. 

The good news is that the open data are now available at 0.1 degree resolution and I believe (but have not confirmed) it is accessible without any Python wrappers via the [ecmwfr](https://cran.r-project.org/web/packages/ecmwfr/index.html) package, so a replacement function should be more straightforward than was possible in 2022.

@istfer does this make sense to you?

istfer commented:
Agreed @infotroph please go ahead and close this, thank you
